### PR TITLE
feat: add attachment with multer [backend]

### PR DESCRIPTION
**Pull request by Arssin**
- Not sure if md5 hex (32signs) is ok, (on attachment init is 24  signs, but this two are hardcoded - probably gonna be deleted anyway in later steps of project)
- Multer working great, saving files in database - tested on Postman
- Adding two files with same name generating diffrent fileNameHashes, both are working( in my opinion it's much better then being allow to add 1 hash - 1 name)

### CODE REVIEW

**Comment by Arssin** [comment]
- Added delete and get in attachments controllers, added routes and postman.
- @mcwiekala probably will be feedback with DTO, but 1st, I need to read and see examples about it.

File: postman/camp-trello-clone.postman_attachments.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{
```
Comment (mcwiekala) [inline postman/camp-trello-clone.postman_attachments.json:1]: add this collection as a folder to existing collection. Don't create another one

File: packages/backend/src/app.ts [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
 import mongoSanitize from 'express-mongo-sanitize'
+import path from 'path'
 import cors from 'cors'
-import router from './infrastructure/express/router'
 
-const app = express()
+const app: express.Application = express()
 app.use(mongoSanitize())
 
+app.use('../uploads', express.static(path.join(__dirname, 'uploads')))
```
Comment (mcwiekala) [inline packages/backend/src/app.ts:9]: what is the reason to do that here? Is it used at all? Why its not in `attachment.routes.ts`?

File: packages/backend/src/helpers/attachmentsHelper.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import multer from 'multer'
```
Comment (mcwiekala) [inline packages/backend/src/helpers/attachmentsHelper.ts:1]: the logic like this should be in: attachment.service.ts
There is no need to create separate file for it

File: packages/backend/src/helpers/fileNameHash.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import crypto from 'crypto'
```
Comment (mcwiekala) [inline packages/backend/src/helpers/fileNameHash.ts:1]: the logic like this should be in: attachment.service.ts
There is no need to create separate file for it

File: packages/backend/src/infrastructure/mongoose/index.ts [hunk 5]
```diff
@@ -1,11 +1,14 @@
...
+import mongoose, { ConnectOptions } from 'mongoose'
 
 export const connectToDatabase = async () => {
   const { DB_USERNAME, DB_PASSWORD } = process.env
   const dbURL = `mongodb://${DB_USERNAME}:${DB_PASSWORD}@localhost:27017/trello-clone`
 
   try {
-    await mongoose.connect(dbURL)
+    await mongoose.connect(dbURL, {
+      useNewUrlParser: true,
```
Comment (mcwiekala) [inline packages/backend/src/infrastructure/mongoose/index.ts:9]: (optional) it looks like its not necessary
https://mongoosejs.com/docs/migrating_to_6.html#no-more-deprecation-warning-options

Comment (mcwiekala) [inline packages/backend/src/infrastructure/mongoose/index.ts:9]: (optional) it looks like its not necessary
https://mongoosejs.com/docs/migrating_to_6.html#no-more-deprecation-warning-options

File: packages/backend/src/modules/attachment/Attachment.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+export class Attachment {
+  private _id: string
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/Attachment.ts:2]: final?

File: packages/backend/src/modules/attachment/attachment.repository.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import fsPromises from 'fs/promises'
+import { CreateAttachmentDTO, AttachmentDTO } from 'shared'
+import { Attachment } from './Attachment'
+import { fileNameHash } from '../../helpers/fileNameHash'
+import { Repository } from '../../application/Repository'
+import AttachmentModel from './attachments.model'
+
+export class AttachmentRepository implements Repository<AttachmentDTO> {
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachment.repository.ts:8]: DTO should be never stored in repository. Why you created mapper if at the end of the day the dto is stored?

Comment (mcwiekala) [reply packages/backend/src/modules/attachment/attachment.repository.ts:8]: please add correct methods to this Repository interface (save/findById/deleteById)

File: packages/backend/src/modules/attachment/Attachment.js [hunk 8]
```diff
@@ -4,9 +4,9 @@ const { ObjectId } = mongoose.Schema
 
 const attachmentSchema = new mongoose.Schema({
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/Attachment.js:5]: rename file to: attachment.model.ts

File: packages/backend/src/modules/attachment/Attachment.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+export class Attachment {
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/Attachment.ts:1]: rename file to attachment.ts

File: packages/backend/src/modules/attachment/attachments.controller.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import express from 'express'
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachments.controller.ts:1]: typo in filename: attachment

File: packages/backend/src/modules/attachment/attachments.model.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import mongoose from 'mongoose'
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachments.model.ts:1]: typo in filename: attachment

Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachments.model.ts:1]: typo in filename: attachment

Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachments.model.ts:1]: typo in filename: attachment

File: postman/camp-trello-clone.postman_attachments.json [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  "info": {
+    "_postman_id": "c7d3e5c5-dc3b-436c-ac3f-2bc6f3525fc1",
+    "name": "camp-trello-clone",
+    "schema": "https://schema.getpostman.com/json/collection/v2.0.0/collection.json"
+  },
+  "item": [
+    {
+      "name": "POST Attachment",
```
Comment (mcwiekala) [inline postman/camp-trello-clone.postman_attachments.json:9]: give meaningful names to commands eg: create attachement

File: postman/camp-trello-clone.postman_attachments.json [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        "body": {
+          "mode": "formdata",
+          "formdata": [{ "key": "file", "type": "file", "src": "test.jpg" }]
+        },
+        "url": "http://localhost:8085/v1/attachments/"
+      },
+      "response": []
+    },
+    {
+      "name": "GET Attachment",
```
Comment (mcwiekala) [inline postman/camp-trello-clone.postman_attachments.json:23]: give meaningful names to commands

Comment (mcwiekala) [inline postman/camp-trello-clone.postman_attachments.json:23]: give meaningful names to commands

File: postman/camp-trello-clone.postman_attachments.json [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+          "host": ["localhost"],
+          "port": "8085",
+          "path": ["v1", "attachments", ":attachmentId"],
+          "variable": [{ "key": "attachmentId", "value": null }]
+        }
+      },
+      "response": []
+    },
+    {
+      "name": "DEL Attachment",
```
Comment (mcwiekala) [inline postman/camp-trello-clone.postman_attachments.json:40]: give meaningful names to commands

File: packages/backend/src/modules/attachment/attachment.service.ts [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+import { AttachmentRepository, attachmentRepository } from './attachment.repository'
+
+export class AttachmentService {
+  private readonly _attachmentRepository: AttachmentRepository
+
+  constructor(attachmentRepository: AttachmentRepository) {
+    this._attachmentRepository = attachmentRepository
+  }
+
+  async create(dto: CreateAttachmentDTO): Promise<Attachment> {
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachment.service.ts:12]: is that really a dto in parameter? Isn't it just a file?

File: packages/backend/src/modules/attachment/attachments.controller.ts [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+class AttachmentsController {
+  private readonly _attachmentMapper: AttachmentMapper
+  private readonly _attachmentService: AttachmentService
+
+  constructor(attachmentService: AttachmentService, attachmentMapper: AttachmentMapper) {
+    this._attachmentMapper = attachmentMapper
+    this._attachmentService = attachmentService
+  }
+
+  async create(req: any, res: express.Response) {
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachments.controller.ts:16]: remove this any and use express.Request :p

File: postman/camp-trello-clone.postman_attachments.json (hunk 1 above)
Comment (mcwiekala) [inline postman/camp-trello-clone.postman_attachments.json:1]: use PORT variable in postman URL: http://localhost:{{PORT}}
we use 8800 not 8085

File: packages/backend/src/modules/attachment/attachments.controller.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+    const createdAttachment: Attachment = await this._attachmentService.create(req.file)
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachments.controller.ts:17]: we need taskId to know to which task this attachement belongs use:
`interface CreateAttachmentDTO {
  taskId: string
  file: File
}`

Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachments.controller.ts:17]: we need taskId to know to which task this attachement belongs use:
`interface CreateAttachmentDTO {
  taskId: string
  file: File
}`

**Comment by Arssin** [comment]
@mcwiekala Done

File: packages/backend/src/modules/attachment/attachment.repository.ts (hunk 7 above)
Comment (mcwiekala) [reply packages/backend/src/modules/attachment/attachment.repository.ts]: fix it

File: packages/backend/src/modules/attachment/Attachment.ts (hunk 9 above)
Comment (mcwiekala) [reply packages/backend/src/modules/attachment/Attachment.ts:1]: fix it

File: packages/backend/src/modules/attachment/attachment.controller.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import express from 'express'
+import { AttachmentDTO, CreateAttachmentDTO } from 'shared'
+import { Attachment } from './attachment'
+import { AttachmentService, attachmentService } from './attachment.service'
+import { AttachmentMapper, attachmentMapper } from './attachment.mapper'
+
+class AttachmentsController {
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachment.controller.ts:7]: add new method downloadFile where in request there is fileId, and in the response there is file

File: packages/backend/src/modules/attachment/attachment.controller.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    const savedAttachmentDto: AttachmentDTO = this._attachmentMapper.mapToDto(savedAttachment)
+    return res.status(201).json(savedAttachmentDto)
+  }
+
+  async findById(req: express.Request, res: express.Response) {
+    const singleAttachment: Attachment = await this._attachmentService.findById(
+      req.params.attachmentId
+    )
+    const dto: AttachmentDTO = this._attachmentMapper.mapToDto(singleAttachment)
+    return res.status(200).json(dto)
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachment.controller.ts:31]: status code

File: packages/backend/src/modules/attachment/attachment.controller.ts [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    const createdAttachmentDto: CreateAttachmentDTO = { taskId: req.body.taskId, file: req.file }
+
+    const savedAttachment: Attachment = await this._attachmentService.createAttachment(
+      createdAttachmentDto
+    )
+    const savedAttachmentDto: AttachmentDTO = this._attachmentMapper.mapToDto(savedAttachment)
+    return res.status(201).json(savedAttachmentDto)
+  }
+
+  async findById(req: express.Request, res: express.Response) {
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachment.controller.ts:26]: use try catch to handle errors. Right now when the attachementId is wrong the whole app is crashing. Thats definetly what we want do avoid. 404 is ok for error like this

File: packages/backend/src/modules/attachment/attachment.repository.ts [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  async findById(id: string): Promise<Attachment> {
+    const attachment: Attachment = await this._attachmentModel.findById(id)
+    return attachment
+  }
+
+  async deleteById(id: string): Promise<Attachment> {
+    const file = await this._attachmentModel.findById(id)
+    // TODO: this method shouldn't return null
```
Comment (mcwiekala) [inline packages/backend/src/modules/attachment/attachment.repository.ts:35]: this method shouldn't return null please fix

**Review by mcwiekala — APPROVED** [review]
cool!

**Pull request merged**